rust/src/pgsql/pgsql.rs

```diff
@@ -0,0 +1,879 @@
+        if self.request_gap {
+            if !probe_ts(input) {
+                // The parser now needs to decide what to do as we are not in sync.
+                // For now, we'll just try again next time.
+                SCLogDebug!("Suricata interprets there's a gap in the request");
+                return AppLayerResult::ok();
+            }
+
+            // It looks like we're in sync with the message header
+            // clear gap state and keep parsing.
+            self.request_gap = false;
+        }
+
+        let mut start = input;
+        while start.len() > 0 {
+            SCLogDebug!(
+                "In 'parse_request' State Progress is: {:?}",
+                &self.state_progress
+            );
+            match PgsqlState::state_based_req_parsing(self.state_progress, start) {
+                Ok((rem, request)) => {
+                    start = rem;
+                    if let Some(state) = PgsqlState::request_next_state(&request) {
+                        self.state_progress = state;
+                    };
+                    let tx_completed = self.is_tx_completed();
+                    if let Some(tx) = self.find_or_create_tx() {
+                        tx.request = Some(request);
+                        if tx_completed {
+                            tx.tx_state = PgsqlTransactionState::ResponseDone;
+                        }
+                    } else {
+                        // If there isn't a new transaction, we'll consider Suri should move on
+                        return AppLayerResult::ok();
+                    };
+                }
+                Err(nom::Err::Incomplete(_needed)) => {
+                    let consumed = input.len() - start.len();
+                    let needed_estimation = start.len() + 1;
+                    SCLogDebug!(
+                        "Needed: {:?}, estimated needed: {:?}",
+                        _needed,
+                        needed_estimation
+                    );
+                    return AppLayerResult::incomplete(consumed as u32, needed_estimation as u32);
+                }
+                Err(_) => {
+                    return AppLayerResult::err();
+                }
+            }
+        }
+
+        // Input was fully consumed.
+        return AppLayerResult::ok();
+    }
+
+    /// When the state changes based on a specific response, there are other actions we may need to perform
+    ///
+    /// If there is data from the backend message that Suri should store separately in the State or
+    /// Transaction, that is also done here
+    fn response_process_next_state(
+        &mut self, response: &PgsqlBEMessage, f: *const Flow,
+    ) -> Option<PgsqlStateProgress> {
+        match response {
+            PgsqlBEMessage::SSLResponse(parser::SSLResponseMessage::SSLAccepted) => {
+                SCLogDebug!("SSL Request accepted");
+                unsafe {
+                    AppLayerRequestProtocolTLSUpgrade(f);
```

Comment:
Should this be changed to use our rust TLS parser?

Reply:
No all this does it tell the app layer api to see if the flow can switch to tls, so no need to worry about handling the tls itself.

Reply:
oki, thanks!